### PR TITLE
Fix/1743

### PR DESCRIPTION
**Pull request by davilera**
This PR addresses issue #1743. It adds a new renderer setting named `add_root_updir` to let the user customize whether `/..` should be appended (or not) to the root directory.

### CODE REVIEW

**Comment by gegoune** [comment]
Perhaps instead of adding another narrow setting we could allow users to set `header` (or similar) to either string or function? Some data could be passed to callback. This would allow greater flexibility without adding various toggles.

@alex-courtis What do you think? There are most likely other parameters that could benefit from more generalised solution.

**Comment by petobens** [comment]
> Perhaps instead of adding another narrow setting we could allow users to set header (or similar) to either string or function? Some data could be passed to callback. This would allow greater flexibility without adding various toggles.

Big plus 1 to this. I asked for something similar here https://github.com/nvim-tree/nvim-tree.lua/issues/876#issuecomment-1030631944

**Comment by alex-courtis** [comment]
> Perhaps instead of adding another narrow setting we could allow users to set `header` (or similar) to either string or function? Some data could be passed to callback. This would allow greater flexibility without adding various toggles.
> 
> @alex-courtis What do you think? There are most likely other parameters that could benefit from more generalised solution.

Yes. Such flexibility via functions desirable. We do this with `sort_by` `view.width` `view.float.open_win_config` `on_attach`

Let's continue with this pattern.

Proposed help:

```
    *nvim-tree.renderer.root_folder_modifier*
    In what format to show root folder. `"/.."` will be appended.
    See |filename-modifiers| for available options.
      Type: `string`, Default: `":~"`

    *nvim-tree.renderer.root_folder_name*
    Function to set the root folder's displayed name.
    When not set, the root folder title is |nvim-tree.renderer.root_folder_modifier|
      Type: `function`, Default: `nil`

      Function is passed the absolute path of the root folder and should return a string.
      e.g. >
        root_folder_title = function(path)
          return "root directory: " .. vim.fn.fnamemodify(path)
        end
<
```

**Comment by alex-courtis** [comment]
Are you comfortable implementing this @davilera ? It looks like most of the parts are there.

**Comment by gegoune** [comment]
Would we need both? Can't `nvim-tree.renderer.root_folder` take `function` or `string`? If `string` we can run it through `fnamemodify`, if `function` it will be up to user to return what they want. Can still default to `":~"`

**Comment by alex-courtis** [comment]
> Would we need both? Can't `nvim-tree.renderer.root_folder` take `function` or `string`? If `string` we can run it through `fnamemodify`, if `function` it will be up to user to return what they want. Can still default to `":~"`

Of course! That is a much better experience.

We could then "silently move": `root_folder_modifier` -> `root_folder`.

**Comment by gegoune** [comment]
Although maybe `root_folder_label` is a better name?

**Comment by davilera** [comment]
> Are you comfortable implementing this @davilera ?

I'll be happy to give it a shot! I'll work on the PR later on today :-)

> Although maybe `root_folder_label` is a better name?

Ok! Just to be clear, in my PR I'll get rid of `root_folder_modifier` (as well as the one I proposed: `add_root_updir`) and I'll add `root_folder`.

**Comment by davilera** [comment]
There you go, guys. By the way, I renamed the setting to `root_folder_label` as proposed by @gegoune... that's what I wanted to say in my previous message, but I made a mistake.

If someone could please review the commit, I'd really appreciate it.

**Comment by gegoune** [comment]
LGTM! Thank you so much!

@alex-courtis I have added warning for this breaking change, also moved `opts.focus_empty_on_setup = nil` to conditional statement, I am not sure if it wasn't in there by mistake or not. Do you mind having a look?

**Comment by alex-courtis** [comment]
> @alex-courtis I have added warning for this breaking change, also moved `opts.focus_empty_on_setup = nil` to conditional statement, I am not sure if it wasn't in there by mistake or not. Do you mind having a look?

Good catch.

I moved the change from removed to refactored, so that there's no warning, if that's OK with you.

File: lua/nvim-tree/renderer/builder.lua
Comment (alex-courtis) [inline lua/nvim-tree/renderer/builder.lua:299]: We have lost the `/..` at the end of the default which might upset some users.

Proposal: append it to the default: `root_folder_label = ":~:s?$?/..?"`

That adds some complexity. I'd appreciate some test/feedback.

**Comment by gegoune** [comment]
That makes more sense, thanks.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Tested function OK:
* empty string
* populated string

Fail:
* return number
* return table

**Comment by alex-courtis** [comment]
> That makes more sense, thanks.

I'm not sure what to do about those ones... they'll hang around forever...

File: lua/nvim-tree/renderer/builder.lua
Comment (gegoune) [reply lua/nvim-tree/renderer/builder.lua:299]: Doesn't matter how complex default is. Most users won't ever change it.

**Comment by gegoune** [comment]
@davilera Can you enforce return type of that function to string? It should be done after that I think.

**Comment by davilera** [comment]
@gegoune, I just pushed a couple of commits to address this:

> enforce return type of that function to string

and this:

> We have lost the `/..` at the end of the default which might upset some users. Proposal: append it to the default

I hope it works as expected!

File: lua/nvim-tree/renderer/builder.lua
Comment (gegoune) [inline lua/nvim-tree/renderer/builder.lua:302]: Thanks, looks good, I am only thinking whether we should notify user that returned value must be string and until it is default value will be used instead. `vim.notify_once()`?

**Comment by gegoune** [comment]
Looks good to me, thank you! I have left one more comment above and will leave it to @alex-courtis to give it one last look before merging it. Awesome work!

File: lua/nvim-tree/renderer/builder.lua
Comment (davilera) [reply lua/nvim-tree/renderer/builder.lua:302]: I just realized I forgot to add make the `label` variable `local` in line 301. Should I commit the fix?

**Comment by alex-courtis** [comment]
> An idea for most likely follow up PR. How often will that function run? It could be used to display other things such as summary of git or diagnostic statuses. Eg: `myroot/proj [E6,W3,H9]`.

Yes. We'll need to experiment.

I did cowboy an API addition `get_nodes` https://github.com/nvim-tree/nvim-tree.lua/commit/e38e061710c593d9b88c8ebb2c48f2842c89dc31 however that's minimal and missing things like diagnostics.

File: lua/nvim-tree/renderer/builder.lua
Comment (alex-courtis) [inline lua/nvim-tree/renderer/builder.lua:305]: nice

**Comment by alex-courtis** [comment]
Thank you all... I'm using this now.

**Pull request merged**
